[Outlook](NAA) add testing guidance to FAQ (#4997)
OfficeDev/office-js-docs-pr: docs/outlook/faq-nested-app-auth-outlook-legacy-tokens.md

};
```

> ### How do I test my update add-in

Reviewer: ```suggestion
### Test your updated add-in
```

Reviewer: Use second person pronouns, as if you are speaking to the customer.

> Once you've updated your add-in to use NAA you should test it on all platforms you support, such as Mac, mobile, web, and Outlook on Windows.

Reviewer: ```suggestion
Once you've updated your add-in to use NAA, you should test it on all platforms you support, such as Mac, mobile, web, and Outlook on Windows.
```

#### Test when Exchange tokens turned off

> Test that your add-in works correctly when Exchange tokens are turned off. To do this, deploy your add-in to a tenant with tokens turned off and test it. To turn tokens off, see [Turn legacy Exchange Online tokens on or off](turn-exchange-tokens-on-off.md).

Reviewer: ```suggestion
To test that your add-in works correctly when Exchange tokens are turned off, deploy your add-in to a tenant with tokens turned off and test it. To turn tokens off, see [Turn legacy Exchange Online tokens on or off](turn-exchange-tokens-on-off.md).
```

> If you've implemented a pattern where your code uses Exchange tokens, but then falls over if they are unavailable, be sure you are checking for the correct errors. When a call to get an Exchange token fails, check the [asyncResult.diagnostics](/javascript/api/office/office.asyncresult). If either of the following errors is returned, switch to NAA.

Reviewer: ```suggestion
If you've implemented a pattern where your code uses Exchange tokens but then falls over if they are unavailable, be sure you are checking for the correct errors. When a call to get an Exchange token fails, check the [asyncResult.diagnostics](/javascript/api/office/office.asyncresult). If either of the following errors is returned, switch to NAA.
```

};
```

### Test your updated add-in

Once you've updated your add-in to use NAA, you should test it on all platforms you support, such as Mac, mobile, web, and Outlook on Windows.

#### Test when Exchange tokens turned off

To test that your add-in works correctly when Exchange tokens are turned off, deploy your add-in to a tenant with tokens turned off and test it. To turn tokens off, see [Turn legacy Exchange Online tokens on or off](turn-exchange-tokens-on-off.md).

If you've implemented a pattern where your code uses Exchange tokens but then falls over if they are unavailable, be sure you are checking for the correct errors. When a call to get an Exchange token fails, check the [asyncResult.diagnostics](/javascript/api/office/office.asyncresult). If either of the following errors is returned, switch to NAA.

- `GenericTokenError: An internal error has occurred.`
- `InternalServerError: The Exchange server returned an error. Please look at the diagnostics object for more information.`

#### Test fallback code for Outlook on Windows "about:blank" bug

> Outlook 2016 and 2019 on Windows have a bug when using WebView2. MSAL v3 will detect that NAA is not available and attempt to open a dialog from the task pane to sign in the user. MSAL calls `window.open` with `about:blank`, but Outlook does not handle this correctly and displays an error to the user.

Reviewer 2: WebView2 supports about:blank, it is IE that does not support it

> Outlook 2016 and 2019 on Windows have a bug when using WebView2. MSAL v3 will detect that NAA is not available and attempt to open a dialog from the task pane to sign in the user. MSAL calls `window.open` with `about:blank`, but Outlook does not handle this correctly and displays an error to the user.

Author: ```suggestion
Outlook 2016 and 2019 on Windows have a bug when using the Trident+ (Internet Explorer 11) webview. MSAL v3 will detect that NAA is not available and attempt to open a dialog from the task pane to sign in the user. MSAL calls `window.open` with `about:blank`, but the webview does not handle this correctly and displays an error to the user.
```

> We provided sample code to work around this by using the Office dialog API to successfully sign in the user. For more information, see [Outlook add-in with SSO using nested app authentication](https://github.com/OfficeDev/Office-Add-in-samples/tree/main/Samples/auth/Outlook-Add-in-SSO-NAA), and specifically the **Fallback code** section of the readme.

Reviewer 2: You might want to reference the IE sample, since this bug only applies to IE.
